### ISSUE TITLE
Tuple destructuring is broken with new tuple syntax

### ISSUE THREAD

**Issue by SupunS** [ballerina-platform/ballerina-lang#16029]
**Description:**

Previous syntax that works:
```ballerina
public function main(string... args) {
    var (a, (b, c)) = ("Ballerina", (123, true));
        ((a, b), c) = (("UpdatedBallerina", 453), false);
}
```
New syntax:
```ballerina
public function main(string... args) {
    var [a, [b, c]] = ["Ballerina", [123, true]];
    [[a, b], c] = [["UpdatedBallerina", 453], false];
}
```
But get the below errors:
```
error: .::testMain.bal:3:13: invalid tuple variable; expecting a tuple type but found 'int|boolean[]' in type definition
error: .::testMain.bal:4:10: undefined symbol 'b'
error: .::testMain.bal:4:14: undefined symbol 'c'
```

**Comment by grainier** [comment]
According to the spec [1], "If there is no contextually expected type, then the inherent type will be T[], where T is the union of the broad types of the expressions in expr-list. It is an compile-time error if expr-list is empty and there is no contextually expected type." Since, this is defined with var, and no contextually expected type, [123, true] will get inferred as (int|boolean)[], and therefore ["Ballerina", [123, true]] as (string|(int|boolean)[])[]. Therefore, I think we cannot support the above. 

@hasithaa, @sameerajayasoma WDYT?

[1] https://github.com/ballerina-platform/ballerina-spec/blob/master/lang/spec.html#L2712

**Comment by hasithaa** [comment]
Yes, above doesn't work. The compiler can't infer the type from, the LHS, So it tries to infer the RHS expression type from the array literal. Best it can do is, infer the RHS side type as `(string|(int|boolean)[])[]`.
Anyway, if either side expressions have the contextual type, this problem won't occur. Anyway, it is weird to see someone writes something like this.  
`var [a, [b, c]] = ["Ballerina", [123, true]];`

**Comment by SupunS** [comment]
I have a different opinion :)

Here LHS is a tuple destructure, and hence the context sort of knows  the expected type/shape. Even now, it seems the compiler have correctly identified `[a, [b, c]]` as a tuple and tried to infer the type of its members: `a` and `[b, c]`.

I think what the spec refers to as "no contextually typed" is scenarios such as `var x = ["Ballerina", [123, true]]`, `any x = ["Ballerina", [123, true]]`, etc.

**Comment by hasithaa** [comment]
@SupunS , Shall we move this discussion to spec repo?
